[KV Connector][Metrics] Do not count local prefix cache hits in connector queries

markmc opened the pull request:
Somewhat related to #28585

The `vllm:external_prefix_cache_queries` metric was double-counting queries by recording the total prompt tokens instead of only the tokens actually queried from the KV connector.

Example scenario:
- Request with 1000 prompt tokens
- Local cache finds 600 tokens
- External cache finds 200 of the remaining 400 tokens
- Computed: 200 tokens

Metrics before this fix:
```
vllm:prefix_cache_queries: 1000
vllm:prefix_cache_hits: 600
vllm:external_prefix_cache_queries: 1000 # Double counting!
vllm:external_prefix_cache_hits: 200
```

We introduce a slight change in behavior - KV connector "hits" are counted, even if the load subsequently fails. Load failure metrics. Failed transfers can be captured by metrics like `vllm:nixl_num_failed_transfers` and the `vllm:prompt_tokens_by_source_total` in #33289 will arguably will offer better insight into the breakdown of where prompt KV was computed.

markmc committed:
[KV Connector][Metrics] Do not count prefix cache hits in connector queries

Somewhat related to #28585

The `vllm:external_prefix_cache_queries` metric was double-counting
queries by recording the total prompt tokens instead of only the
tokens actually queried from the KV connector.

Delay the recording of connector prefix cache queries and hits until
after KV load has succeeded or failed. In the sync connector mode,
we only know a KV load has succeeded after the model step has
completed.

So:

- We record the queried/hit token count on the request
- In update_from_output(), we record these stats from successful
  requests, and include them in the SchedulerStats for this
  iteration
- If a reset comes in, we note this so that it can also be
  included in the next SchedulerStats

Example scenario:
- Request with 1000 prompt tokens
- Local cache finds 600 tokens
- External cache finds 200 of the remaining 400 tokens
- Computed: 200 tokens

Metrics before this fix:
```
vllm:prefix_cache_queries: 1000
vllm:prefix_cache_hits: 600
vllm:external_prefix_cache_queries: 1000 # Double counting!
vllm:external_prefix_cache_hits: 200
```

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -1722,7 +1735,10 @@ def _update_requests_with_invalid_blocks(
+                if request.connector_prefix_cache_queries:
+                    request.connector_prefix_cache_queries -= num_affected_tokens
+                if request.connector_prefix_cache_hits:
+                    request.connector_prefix_cache_hits -= num_affected_tokens
```
This comment seems correct for the recompute case, but not if we fail the request. Will dig some more

markmc replied:
Ok, this should be fixed now - we don't record the queries/hits at all for failed requests, and we only record the queries for recomputed requests

orozery commented:
I don't think we want to hide the original number of hits (subtracting the errored one as proposed here).
Instead, we can introduce something like `external_prefix_cache_errors`.
We can observe this new metric without introducing new state to `Request`, which I think is preferrable.

markmc committed:
[KV Connector][Metrics] Remove queries/hits from Request state

Rather than trying to ensure that a failed KV load does not count
as a "hit", let's simplify and reduce runtime state by ignoring
these failures in the connector prefix cache metrics.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

orozery reviewed (approved):
Looks great!

markmc committed:
[KV Connector][Metrics] Reinstate Request.num_external_computed_tokens

Needed by #33290

Signed-off-by: Mark McLoughlin <markmc@redhat.com>
